Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -621,9 +621,66 @@ However, this is not a concern on iOS. As discussed in the section [Testing on t
+stp        x22, x21, [sp, #-0x30]!
+mov        rbp, rsp
+```
+
+After the obfuscation we can observe that the symbol’s name is no longer meaningful as shown on the listing below.
+
+```assembly
+__T07DVIA_v232zNNtWKQptikYUBNBgfFVMjSkvRdhhnbyyFySbyypF:
+stp        x22, x21, [sp, #-0x30]!
+mov        rbp, rsp
+```
+
+Nevertheless, the binary’s code is not modified, so an attacker can still retrieve the logic of the algorithm by reading the disassembled version of the function and figure out its purpose.
+
+#### Instructions substitution
+
+This technique replaces standard binary operators like addition, subtraction with more complex representations. For example an addition `x = a + b` can be represented as `x = -(-a) - (-b)`. However, using the same replacement representation could be easily reversed, so it is recommended to add multiple substitution techniques for a single case and introduce a random factor. Nevertheless, this technique is vulnerable to code re-optimaliztion.
+
+#### Control flow Flattering
+Control flow Flattening replaces original code with a more complex representation. The transformation breaks the body of the function into basic blocks and puts them all inside a single infinite loop with a switch statement that controls the program flow. This makes the program flow significantly harder to follow because the natural conditional constructs that made the code easier to read are now gone.
+
+
+
+![control-flow-flattening](./Images/Chapters/0x06j/control-flow-flattening.png)
+
+The image shows how the control-flow flattening alters code ([Obfuscating C++ programs via control flow flattening](http://ac.inf.elte.hu/Vol_030_2009/003.pdf))
+
+#### Dead code injection
+
+This technique makes the program's control-flow more complex by injecting dead code into the program. Dead code is a stub of code that doesn’t affect the original program’s behaviour but brings the overhead for the reverse engineering.
+
+#### String encryption
+
+Applications are often compiled with hardcoded keys, licences, tokens and URL endpoints. By default, all of them are stored in plaintext in the *data section* of an application’s binary. String encryption aims to replace these values with their encrypted forms and inject stubs of code into the program that decrypts data before it is used by the program. 
+
+#### Recommended tools
+
+* [SwiftShield](https://github.com/rockbruno/swiftshield) can be used to obfuscate binary's symbols. It reads the source code of the XCODE project and replaces all names of classes, methods and fields with random values before the compiler is used.
+* [obfuscator-llvm](https://github.com/obfuscator-llvm) operates on the Intermediate Representation instead of the the source code. It can be used for symbols obfuscation, string encryption and control-flow flattening. Since it's based on IR, it can hide out significantly more information about the application as compared to SwiftShield.
```

Comment:
```suggestion
* [obfuscator-llvm](https://github.com/obfuscator-llvm/obfuscator/tree/llvm-4.0 "obfuscator-llvm) operates on the Intermediate Representation (IR) instead of the the source code. It can be used for symbols obfuscation, string encryption and control-flow flattening. Since it's based on IR, it can hide out significantly more information about the application as compared to SwiftShield.
```

Did you try out obfuscator-llvm? This project hasn't been updated for 3. years.

Reply:
Hey @sushi2k. Thanks for the review! I haven't tested _obfuscator-llvm_ myself but I rather wanted to point out a compiler based solutions here. It seems to work based on XODE plugins which are no longer supported in newer versions of XCODE but there is a [Hikari](https://github.com/HikariObfuscator/Hikari) that does pretty similar job using  xcode-toolchain. Should we add a sentence about it?

Reply:
Hey @serek8. Noted and thanks for pointing out Hikari.  Yes, please mention it also, it is actively maintained.

Reply:
After adding, we are good to go from my side. Thanks for the enhancement, really good work!